### ISSUE TITLE
Loopback API won't response after mongodb connection fail.

### ISSUE THREAD

**Issue by fhsakai** [strongloop/loopback-connector-mongodb#333]
Hello,

I'm facing a problem with loopback API using Model connect to  MongoDB (V3.4).
When the DB is down or the network connection to DB fail for more than 1 minute with default connection configuration the API stop to work properly even when DB became online after that time.

All request return:
```
{
  "error": {
    "statusCode": 500,
    "message": "Internal Server Error"
  }
}
```

and in the server log we can see the error:

> MongoError: Topology was destroyed

As I couldn't find a solution, would like to ask for assistance.

### Scenario 1 

1. MongoDB is online
2. Loopback API is online
3. Loopback API call responses is OK 
4. Stop call Loopback API
5. Stop MongoDB or emulate a network fail between API server and MongoDB
6. Wait more then 1 minute
7. Start MongoDB
8. Execute a call to Loopback API
9. Status Code 500 will return from API
10. API will log "MongoError: Topology was destroyed"
11. All subsequent calls to API will fail with same error

> Solution : Restart Application

### Scenario 2

1. MongoDB is online
2. Loopback API is online
3. Loopback API call responses is OK 
4. Stop MongoDB or emulate a network fail between API server and MongoDB
5. Execute a call to Loopback API
6. Status Code 500 will return from API
7. API will log "MongoError: failed to reconnect after 30 attempts with interval 1000 ms"
8. Start MongoDB
9. Execute a call to Loopback API
10. Status Code 500 will return from API
11. API will log "MongoError: Topology was destroyed"
12. All subsequent calls to API will fail with same error

> Solution : Restart Application

### Expected result
Once the mongoDB is online the Loopback API start to respond properly to the new requests without need to restart the application.

Could you please advice where I'm missing or doing wrong?

Thanks in advanced.

### Datasources configuration
```
{
  "db": {
    "name": "db",
    "connector": "mongodb",
    "host": "192.168.2.62",
    "port": 27017,
    "database": "mydb"
  }
}
```

### Components Version
```

+-- loopback@3.2.1
+-- loopback-connector-mongodb@1.18.0
| +-- async@1.0.0
| +-- loopback-connector@2.7.1
| | `-- msgpack5@3.4.1
| |   `-- bl@1.1.2
| |     `-- readable-stream@2.0.6
| `-- mongodb@2.2.21
|   +-- es6-promise@3.2.1
|   +-- mongodb-core@2.1.6
|   | +-- bson@1.0.4
|   | `-- require_optional@1.0.0
|   |   +-- resolve-from@2.0.0
|   |   `-- semver@5.3.0
|   `-- readable-stream@2.1.5
|     `-- buffer-shims@1.0.0
```

**Comment by smartmouse** [comment]
Have you checked the connection properties for mongodb?
http://mongodb.github.io/node-mongodb-native/2.2/reference/connecting/connection-settings/

keepAlive, connectionTimeOut, reconnect, reconnectRetries will all be helpful. You can experiment with them as see if it helps .

**Comment by fhsakai** [comment]
Dear smartmouse,

Thanks for the feedback.

We tried to change those settings before, but the results are same, the difference is only the time when the exception is raised if we increase or decrease the reconnect retries and its interval only the time we need to wait for the errors happens change.
if we change to not reconnect the error happens as soon we execute the call while the db is offline.

**Comment by smartmouse** [comment]
What s your topology and environment? e.g node version, npm version, platform,  your mongodb is remote on another machine?
Have you tried using a very simple sample-demo app, does this also happen?

**Comment by fhsakai** [comment]
Dear @smartmouse ,

Very sorry for late response.
node version : 6.9.1
npm version :  4.0.3

tested platform :
- node : windows 8.1 and mac OSX El Capitan
- mongodb :  : windows 8.1 and mac OSX El Capitan

mongodb connection was tested with:
-  remote connection 
-  local connection

I made  a new test with simple sample as suggested, but the result remain the same as previously reported.
https://github.com/fhsakai/loopback-connector-mongodb-issue-333

Best regards.

**Comment by fhsakai** [comment]
This issue was solved in [v3.1.0](https://github.com/strongloop/loopback-connector-mongodb/releases/tag/v3.1.0)
For details, please refer to : https://github.com/strongloop/loopback-connector-mongodb/pull/362
